ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java

```diff
@@ -108,7 +108,16 @@ public final void check()
                 // TODO close window where first can change - maybe record nextDynamicCheckTime before probing first,
                 // then use compareAndSet? Although something will need to check before sleeping anyhow...
                 if (reaperElement == null) {
-                    nextDynamicCheckTime.set(Long.MAX_VALUE);
+                    if (!_inShutdown)
+                        nextDynamicCheckTime.set(Long.MAX_VALUE);
+                    else {
+                        // If I am here during shutdown it means that _inShutdown was read from a cached value
```

Comment:
@jmfinelli When reading this, it could mean that this thread read the cached value - but I don't think that is the point so maybe `was read by another thread from a cached value`?

Reply:
But also, it probably shouldn't be so assertive because because the other thread might be at https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L735 and about to read _inShutdown as true possibly

Reply:
The solution presented in the branch `JBTM-3830_jfinelli` is an alternative to the [current logic](https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L735). In other words, if we like this alternative better than Steven's PR, I will revert that `wait(1000)` from Narayana's `main` branch

Reply:
Sorry, I think my wording was a bit unclear (though not in the secondary point - as in it may well be possible (due to the lack of synchronization that _inShutdown was about to be read as true in removeElementClient by this other thread.

I will try to work out what I was trying to say in my first comment in this thread :)

Reply:
Ah, I think the first point was more about understanding "who" (which thread) could read the cached value (so it's related to the second point but not the same point). Basically though I think it's correct what you said (except that it doesn't necessary mean it was read from a cached value to be false because of what I said in my second point), it could be helpful to expand the command to make it clear that it would not be _this_ thread that read the cached value. I suppose it might be a bit unnecessary to state that because _this_ thread should not be calling removeElementClient but I think it might still be helpful to the reader. I am going to make a suggested word change to cover both of these points.

Reply:
If the JVM reaches the code then it means that the TransactionReaper is in shutdown and either another thread has called removeElementClient  but read a cached value for _inShutdown, or the other thread is in the process of executing removeElementClient and we preemptively decide to notify here in case _inShutdown returns a stale value of false for that thread.